cmd/githubwebhookserver/metrics/workflow_job_metrics_event_reader.go

```diff
@@ -0,0 +1,133 @@
+package metrics
+
+import (
+	"context"
+	"strings"
+
+	"github.com/go-logr/logr"
+	gogithub "github.com/google/go-github/v45/github"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type WorkflowJobMetricsEventReader struct {
+	Log logr.Logger
+
+	// Event queue
+	Events      chan interface{}
+	trackedJobs map[int64]*gogithub.WorkflowJobEvent
+}
+
+// HandleWorkflowJobEvent send event to reader channel for processing
+//
+// forcing the events through a channel ensures they are processed in sequentially,
+// and prevents any race conditions with githubWorkflowJobStatus
+func (reader *WorkflowJobMetricsEventReader) HandleWorkflowJobEvent(event interface{}) {
+	reader.Events <- event
+}
+
+// ProcessWorkflowJobEvents pop events in a loop for processing
+//
+// Should be called asynchronously with `go`
+func (reader *WorkflowJobMetricsEventReader) ProcessWorkflowJobEvents(ctx context.Context) {
+	if reader.trackedJobs == nil {
+		reader.trackedJobs = make(map[int64]*gogithub.WorkflowJobEvent)
+	}
+	for {
+		select {
+		case event := <-reader.Events:
+			reader.ProcessWorkflowJobEvent(event)
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// ProcessWorkflowJobEvent processes a single event
+//
+// Events should be processed in the same order that Github emits them
+func (reader *WorkflowJobMetricsEventReader) ProcessWorkflowJobEvent(event interface{}) {
+
+	e, ok := event.(*gogithub.WorkflowJobEvent)
+	if !ok {
+		return
+	}
+
+	// collect labels
+	labels := make(prometheus.Labels)
+
+	runsOn := strings.Join(e.WorkflowJob.Labels, `,`)
+	labels["runs_on"] = runsOn
+	labels["job_name"] = *e.WorkflowJob.Name
+
+	job_id := *e.WorkflowJob.ID
+
+	// switch on job status
+	switch action := e.GetAction(); action {
+	case "queued":
+		incrementStatusGauge("queued", labels)
+
+		reader.trackedJobs[job_id] = e
+
+	case "in_progress":
+
+		// discard events if the previous event wasn't tracked
+		previousEvent, ok := reader.trackedJobs[job_id]
```

Comment:
Does this imply that this feature will work only when you have only one replica of the githubwebhookserver, but not for two or more replicas? (It's a standard webhook server so it is expected to be scaled out to 2 or more replicas whenever the operator decides to do so.

Reply:
Yes, that's what I had in mind here. If that's not an assumption I can make, then there needs to be some way to share state, or `github_workflow_job_status` should be dropped

Reply:
@ColinHeathman We may probably introduce an optional dependency to an external distributed kv store to track jobs later. Or GitHub could enhance their API and/or the workflow job event payload so that we can use some information included in the event to measure the queue duration. In the meantime though, we should omit this as this isn't going to work universally

Reply:
I've started a discussion [here](https://github.com/orgs/community/discussions/33640) to enhance the API. I believe that is the cleanest path forwards.

@mumoshu Is there another pending reason to add a kv store? I thought about it but I didn't think these metrics would justify the addition.

Alternatively, we could change `github_workflow_job_status` to a counter, rather than a gauge.
That way queue counts could be calculated by Prometheus, with `github_workflow_job_status{job_status="queued"}[1m] - github_workflow_job_status{job_status="in_progress"}[1m]`

I can't think of a way to make `github_workflow_job_queue` work without a kv store though

Reply:
Thanks for starting the discussion! Let's see how it goes.

> s there another pending reason to add a kv store?

No- this is the first time I thought we might need to add it.

> I thought about it but I didn't think these metrics would justify the addition.

Same- if this isn't a short-term hard requirement, I'd like not to add the kv store as an ARC dependency. Looks too much.

> we could change github_workflow_job_status to a counter, rather than a gauge.

Ah, counts might be good as well! Can we name the metric like `github_workflow_jobs_total(job_status="...")` though? If it's a count, I think that better follows the canonical metric name convention of Prometheus https://prometheus.io/docs/practices/naming/#metric-names